draft-ietf-quic-qcram.md

```diff
@@ -223,15 +331,111 @@ It is an error if the HPACK decoder encounters an indexed representation that
+representation. In particular, when a peer sends a header field that it received
+represented as a literal header field with the 'N' bit set, it MUST use the same
+representation to forward this header field.  This bit is intended for
+protecting header field values that are not to be put at risk by compressing
+them (see Section 7.1 of [RFC7541] for more details).
+
+If the header field name matches the header field name of an entry stored in the
+static table or the dynamic table, the header field name can be represented
+using the index of that entry. In this case, the `S` bit indicates whether the
+reference is to the static (S=1) or dynamic (S=0) table and the index of the
+entry is represented as an integer with an 5-bit prefix (see Section 5.1 of
+[RFC7541]). Valid table indices are always non-zero; a table index of zero
+MUST be treated as a decoding error.
 
-Due to out-of-order arrival, QCRAM's eviction algorithm requires changes
-(relative to HPACK) to avoid the possibility that an indexed representation is
-decoded after the referenced entry has already been evicted.  QCRAM employs a
-two-phase eviction algorithm, in which the encoder will not evict entries that
-have outstanding (unacknowledged) references.
+~~~~~~~~~~
+     0   1   2   3   4   5   6   7
+   +---+---+---+---+---+---+---+---+
+   | 0 | N | S |  Name Index (5+)  |
+   +---+---+-----------------------+
+   | H |     Value Length (7+)     |
+   +---+---------------------------+
+   | Value String (Length octets)  |
+   +-------------------------------+
+~~~~~~~~~~
+{: title="Literal Header Field -- Indexed Name"}
 
-### Blocked Evictions
+Otherwise, the header field name is represented as a string literal (see Section
+5.2 of [RFC7541]). A value 0 is used in place of the 6-bit index, followed by
+the header field name.
+
+~~~~~~~~~~
+     0   1   2   3   4   5   6   7
+   +---+---+---+---+---+---+---+---+
+   | 0 | N |           0           |
+   +---+---+-----------------------+
+   | H |     Name Length (7+)      |
+   +---+---------------------------+
+   |  Name String (Length octets)  |
+   +---+---------------------------+
+   | H |     Value Length (7+)     |
+   +---+---------------------------+
+   | Value String (Length octets)  |
+   +-------------------------------+
+~~~~~~~~~~
+{: title="Literal Header Field -- Literal Name"}
+
+Either form of header field name representation is followed by the header field
+value represented as a string literal (see Section 5.2 of [RFC7541]).
+
+
+# Encoding Strategies
+
+## Reference Tracking
+
+An encoder MUST ensure that a header block which references a dynamic table
+entry is not received by the decoder after the referenced entry has already been
+evicted, and might wish to ensure that the decoder will not suffer head-of-line
+blocking when encoding particular references.
+
+In order to enable this, the encoder MUST track outstanding (unacknowledged)
+header blocks and MAY track outstanding table updates.
```

Comment:
If you explain why each one is tracked, then you might reach the conclusion - as I did - that 2119 language doesn't make sense here.

The encoder tracks outstanding unacknowledged header blocks so that it can avoid causing evictions that might arrive at the decoder before the entries are needed.  The encoder tracks unacknowledged table updates as well so that it can ensure that it never references a table entry that has not been received by the decoder.